presto-main/src/main/java/com/facebook/presto/resourcemanager/ResourceManagerClusterStateProvider.java

```diff
@@ -166,6 +168,37 @@ public void registerNodeHeartbeat(NodeStatus nodeStatus)
         }
     }
 
+    public void registerResourceGroupRuntimeHeartbeat(String node, List<ResourceGroupRuntimeInfo> resourceGroupRuntimeInfos)
+    {
+        resourceGroupStates.put(node, new CoordinatorResourceGroupState(node, resourceGroupRuntimeInfos));
+    }
+
+    public int getAdjustedQueueSize()
+    {
+        Map<ResourceGroupId, ResourceGroupRuntimeInfo.Builder> resourceGroupBuilders = new HashMap<>();
+        resourceGroupStates.values().stream()
+                .map(CoordinatorResourceGroupState::getResourceGroups)
+                .flatMap(Collection::stream)
+                .forEach(resourceGroupRuntimeInfo -> {
+                    ResourceGroupId resourceGroupId = resourceGroupRuntimeInfo.getResourceGroupId();
+                    ResourceGroupRuntimeInfo.Builder runtimeInfoBuilder = resourceGroupBuilders.computeIfAbsent(resourceGroupId, ResourceGroupRuntimeInfo::builder);
+                    runtimeInfoBuilder.addQueuedQueries(resourceGroupRuntimeInfo.getQueuedQueries());
+                    runtimeInfoBuilder.addRunningQueries(resourceGroupRuntimeInfo.getRunningQueries());
+                    runtimeInfoBuilder.addDescendantQueuedQueries(resourceGroupRuntimeInfo.getDescendantQueuedQueries());
+                    runtimeInfoBuilder.addDescendantRunningQueries(resourceGroupRuntimeInfo.getDescendantRunningQueries());
+                    if (resourceGroupRuntimeInfo.getResourceGroupConfigDetails().isPresent()) {
+                        runtimeInfoBuilder.setResourceGroupConfigDetails(resourceGroupRuntimeInfo.getResourceGroupConfigDetails().get());
```

Comment:
may be we can only set it once than updating it multiple times with each coordinator update. May be log in case we get a different spec than the one set for debugging purpose.

Reply:
Based on discussion, moved the computation to scheduledExecutorService